File: scala/DeBruijn.scala
package com.github.antma.cpalgo

class DeBruijn {
  val db = 0x218a392cd3d5dbfL
  val t  = Array.ofDim[Int](64)
  for (k <- 0 to 63) t(((db << k) >>> 58).toInt) = k
  def lookup(x: Long) = t(((x * db) >>> 58).toInt)
}

class DeBruijn32 {
  val db = 0x4653adf
  val t  = Array.ofDim[Int](32)
  for (k <- 0 to 31) t((db << k) >>> 27) = k
  def lookup(x: Int) = t(((x * db) >>> 27).toInt)
}

File: ocaml/debruijn.ml
class debruijn32 =
  object
  val db = Int32.of_string "0x4653ADF"
  val t = Array.make 32 0
  initializer begin
    for k = 0 to 31 do
      t.(Int32.to_int (Int32.shift_right_logical (Int32.shift_left db k) 27)) <- k
    done
  end
  method lookup x = t.(Int32.to_int (Int32.shift_right_logical (Int32.mul db (Int32.of_int x)) 27))
end

let _ = begin
  let db = new debruijn32 in
  for i = 0 to 30 do
    assert (db#lookup (1 lsl i) = i)
  done
end
